Refactor SQL/BigQuery/Qubole/Druid Check operators

xinbinhuang opened the pull request:
<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:


related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

closes: #10271
related: #9844 #14184

This PR refactor SQL/BigQuery Check operators to reduce duplicated code:
- create `BaseSQLOperator`: it standardizes how some of the generic SQL operators retrieve DB hook with the `.get_db_hook()` method
  - Add a `database` kwarg `*CheckOperators` for a consistent interface 
- create `_BigQueryDbHookMixin` to standardize the `.get_db_hook()` method for BigQuery
- create `_QuboleCheckOperatorMixin` to remove duplicate code
- replace `<class-name>.template_fields` with `_get_template_fields` in `__getattribute__` to avoid hard coding class name, and reduce duplicate code
- remove and deprecate `DruidCheckOperator` the same functionality can be achieved by `SQLCheckOperator` - the deprecation method is the same for `PrestoCheckOperator`

Misc:
- Fix docstrings
- Update deprecated Operator name and import path
- Remove unnecessary `if` statements check parameters in `SQLBranchOperator`

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

xinbinhuang committed:
Refactor sql check related operators

xinbinhuang committed:
Refactor bigquery check operators

turbaszek commented on tests/operators/test_sql.py:
```diff
@@ -322,7 +359,7 @@ def test_pass_min_value_max_sql(self, mock_get_db_hook):
         operator.execute()
 
-    @mock.patch.object(ThresholdCheckOperator, "get_db_hook")
+    @mock.patch.object(SQLThresholdCheckOperator, "get_db_hook")
```
Was this name changed?

xinbinhuang replied:
Yeah. The original operators were moved to another module and renamed but the tests were forgotten to refactor I believe.

xinbinhuang replied:
You can see the original imports and the refactored imports at the top of the file - `test_sql.py`

turbaszek replied:
> tests were forgotten to refactor I believe.

That was what I thought, wanted to check. Thanks!

xinbinhuang committed:
fixup! Refactor sql check related operators

xinbinhuang committed:
fixup! fixup! fixup! fixup! fixup! Refactor bigquery check operators

xinbinhuang committed:
Fix tests and typos

xinbinhuang committed:
Fix pylint error

xinbinhuang committed:
Refactor Quobole Operator code

xinbinhuang commented on airflow/providers/qubole/operators/qubole_check.py:
```diff
@@ -16,17 +16,40 @@
+        )
+
+
+# pylint: disable=too-many-ancestors
```
Temporarily disable pylint for these 2 operators. Will submit a second PR in the future to simplify the inheritance.

potiuk replied:
Yeah. for now it is good, but maybe leave a TODO(xinbinhuang)  note as comment?

xinbinhuang committed:
Fix typo

potiuk commented on airflow/operators/sql.py:
```diff
@@ -38,7 +41,53 @@
+        self.log.debug("Get connection for %s", self.conn_id)
+        conn = BaseHook.get_connection(self.conn_id)
+
+        if conn.conn_type not in ALLOWED_CONN_TYPE:
```
Hmm. I think we should rather check the povider's manager hook property for that. We've added automated discovery of providers so that they can register their own connection -> Hook mapping so maybe we should utilise that? https://github.com/apache/airflow/blob/6f246b0d54ccaf733b7c5951a8955adda6719acb/airflow/providers_manager.py#L387

xinbinhuang replied:
I think the reason for `ALLOWED_CONN_TYPE` is to only whitelist a subset of all connection types that are `DbApiHook` compatible. Maybe we can do something like this instead?


```python
hook = conn.get_hook()
if not isinstance(hook, DbApiHook):
    raise AirflowException('The connection type associated hook is not a subclass of `DbApiHook`') 
```

However, there are edge cases that need to be resolved in future PRs to make this work.
- Currently, bigQuery seems to be not discoverable because it overlaps with the GoogleBaseHook: https://github.com/apache/airflow/blob/5f81fc73c8ea3fc1c3b08080f439fa123926f250/airflow/providers/google/cloud/hooks/bigquery.py#L72
- CloudSQLHook/CloudSQLDatabaseHook is not a subclass of `DbApiHook`
https://github.com/apache/airflow/blob/5f81fc73c8ea3fc1c3b08080f439fa123926f250/airflow/providers/google/cloud/hooks/cloud_sql.py#L666
- ... not sure if there are other edge cases

potiuk reviewed (commented):
:heart:  it

Just a few small comment :)

xinbinhuang committed:
Add TODO notes for refactoring inheritance

xinbinhuang committed:
Deprecate DruidCheckOperator

xinbinhuang committed:
Make DruidDbApiHook discoverable by ProvidersManager

fixup! Make DruidDbApiHook discoverable by ProvidersManager

fixup! fixup! Make DruidDbApiHook discoverable by ProvidersManager

xinbinhuang committed:
Rename BigQueryHook conn_type to bigquery

xinbinhuang committed:
Check return hook isinstance of DbApiHook

xinbinhuang committed:
fixup! Deprecate DruidCheckOperator

fixup! fixup! Deprecate DruidCheckOperator

fixup! fixup! fixup! Deprecate DruidCheckOperator

xinbinhuang commented on airflow/providers/apache/druid/operators/druid_check.py:
```diff
@@ -15,74 +15,23 @@
-    :param druid_broker_conn_id: reference to the druid broker
-    :type druid_broker_conn_id: str
+    This class is deprecated.
+    Please use `airflow.operators.sql.SQLCheckOperator`.
```
Is there a deprecation process for Provider packages? i.e. any changes to `provider.yaml` and the README file?

@potiuk

potiuk replied:
There will be very shortly - I am working on adding SemVER versioning changes and procedure for providers.

xinbinhuang replied:
So I will just keep the deprecation inside the `class`  for now. Do you think that is ok?

xinbinhuang committed:
Increae number of provider hooks

2 more hooks are added
- DruidDbApiHook
- BigQueryHook

xinbinhuang committed:
fixup! Deprecate DruidCheckOperator

msumit commented on airflow/providers/qubole/operators/qubole.py:
```diff
@@ -267,7 +267,7 @@ def get_hook(self) -> QuboleHook:
 
     def __getattribute__(self, name: str) -> str:
-        if name in QuboleOperator.template_fields:
+        if name in _get_template_fields(self):
```
what is the rationale behind this change?

xinbinhuang replied:
It avoids hardcoding the class name `QuboleOperator` here, so you can inherit it safely. - so I can remove the `<class-name>.template_fields` in `qubole_check.py`

xinbinhuang committed:
fixup! Increae number of provider hooks

xinbinhuang committed:
Fix sql check tests

turbaszek commented on airflow/operators/sql.py:
```diff
@@ -18,27 +18,59 @@
+            )
+
+        if self.database:
+            hook.schema = self.database
```
This is slightly confusing, database is not the same thing as schema

xinbinhuang replied:
I totally agree!  But this is how the current mapping of Airflow works. Currently, the `Connection` object uses the field `schema` to specify the database (not schema) for database type connections, i.e. Postgres, MySQL, etc. And as a result, all `DbApiHook`s use the field `schema` to specify the database to connect to using the underlying DB API (i.e. `psycopg2`). `extra` is used to specify the schema if applicable. So the relationship looks like this:

```
Connection(schema) -> DbApiHook (schema) -> DB API (dbname/database) 
```

About the field `database`, it was already used for the `SQLBranchOperator` and also other SQL like operators. For this PR, I simply refactor it out to reduce duplicated code and add them to the `check*` operators for consistency.

Though I think it will be nice to consolidate between `schema` and `database`, I think they should be addressed in a separate PR or worth a discussion in the `dev` list. I can send an email in the dev list if you don't object.

xinbinhuang committed:
fixup! fixup! Increae number of provider hooks

xinbinhuang committed:
Replace unittest assertRaise with pytest raises

xinbinhuang committed:
Increase apache.druid minor version to 1.0.2

xinbinhuang committed:
Change to proper naming convention

xinbinhuang committed:
Add DruidCheckOperator to deprecated_classes.py

xinbinhuang commented on tests/deprecated_classes.py:
```diff
@@ -1008,7 +1008,7 @@
     ),
     (
-        'airflow.providers.apache.druid.operators.druid_check.DruidCheckOperator',
+        'airflow.operators.sql.SQLCheckOperator',
```
Deprecate `DruidCheckOperator` in favor of `SQLCheckOperator`

xinbinhuang commented on tests/core/test_core_to_contrib.py:
```diff
@@ -84,7 +84,7 @@ def test_is_class_deprecated(self, new_module, old_module):
                     assert len(warnings) >= 1
                     assert any(warning.filename == __file__ for warning in warnings)
-                init_mock.assert_called_once_with()
+                init_mock.assert_called_once()
```
Not sure if there is a better way to do this while maintaining backward compatibility.  `DruidCheckOperator` has a default `druid_broker_conn_id` that needs to pass down to `SQLCheckOperator.__init__(conn_id=druid_broker_conn_id, **kwargs)`.

So if we do `assert_called_once_with()` the test will fail because `DruidCheckOperator.__ini__()` == `SQLCheckOperator.__init__(conn_id=druid_broker_conn_id)`

turbaszek replied:
@xinbinhuang should this help?
```
assert_called_once_with(conn_id=mock.ANY)
```

xinbinhuang commented on airflow/providers/apache/druid/operators/druid_check.py:
```diff
@@ -15,74 +15,22 @@
-        if not record:
-            raise AirflowException("The query returned None")
-        self.log.info("Success.")
+    def __init__(self, druid_broker_conn_id: str = 'druid_broker_default', **kwargs):
+        warnings.warn(
+            """This class is deprecated.
+            Please use `airflow.operators.sql.SQLCheckOperator`.""",
+            DeprecationWarning,
+            stacklevel=3,
+        )
+        super().__init__(conn_id=druid_broker_conn_id, **kwargs)
```
Passdown `druid_broker_conn_id` to `SQLCheckOperator.__init__`

xinbinhuang reviewed (commented):
@kaxil  Thanks for taking a look! I made a small update to the `DruidCheckOperator` in `deprecated_classes.py` after the original comment. I didn't know that it will break the tests, but now it's fixed.

kaxil commented on airflow/operators/druid_check_operator.py:
```diff
@@ -23,7 +23,7 @@
 
 warnings.warn(
-    "This module is deprecated. Please use `airflow.providers.apache.druid.operators.druid_check`.",
+    "This module is deprecated. Please use `airflow.operators.sql.SQLCheckOperator`.",
```
Oh why is this?

kaxil replied:
nvm saw the change later in the PR

kaxil commented on airflow/providers/google/cloud/hooks/bigquery.py:
```diff
@@ -69,8 +69,8 @@ class BigQueryHook(GoogleBaseHook, DbApiHook):
-    conn_type = 'google_cloud_platform'
-    hook_name = 'Google Cloud'
+    conn_type = 'bigquery'
+    hook_name = 'BigQuery'
```
@turbaszek @potiuk @mik-laj is this change ok?

kaxil replied:
@xinbinhuang Do we need this change?

xinbinhuang replied:
I left a comment about this in the thread previously, but I think it got buried over time. 

This fixes the issue that `conn.get_hook` will never return the bigquery hook after we start to use the `ProvidersManager` because GoogleBaseHook "hides" the BigQueryHook as they have the same conn_type.

Here I copy my original comment:

> Yes, there is one place that may be considered fixing rather than breaking. For airflow/providers/google/cloud/hooks/bigquery.py, I changed the hook's conn_type and hook_name from google cloud platform/Google Cloud to bigquery/BigQuery. This fixes the issue where conn.get_hook will never return the bigquery hook because GoogleBaseHook "hides" the BigQueryHook as they have the same conn_type.

kaxil replied:
Does it needs any action from a user side is why I am thinking about it ? -- if yes, then we might need a note somewhere in Up:

https://github.com/apache/airflow/blob/master/airflow/providers/google/cloud/ADDITIONAL_INFO.md

xinbinhuang replied:
As similar to other Google cloud connection type (before this change, you can't find bigquery in the list)

![image](https://user-images.githubusercontent.com/27927454/109046868-12ecb800-768a-11eb-970c-6d52905825fa.png)
![image](https://user-images.githubusercontent.com/27927454/109046902-1bdd8980-768a-11eb-9cc2-155f26ef7898.png)

Maybe I should rename it to `Google Bigquery` instead?

xinbinhuang replied:
Renamed, this is the new conn_type list for google:
![image](https://user-images.githubusercontent.com/27927454/109047564-de2d3080-768a-11eb-95ee-6e614ae6c27a.png)

xinbinhuang replied:
> Does it needs any action from a user side is why I am thinking about it ? -- if yes, then we might need a note somewhere in Up: https://github.com/apache/airflow/blob/master/airflow/providers/google/cloud/ADDITIONAL_INFO.md

Hmm, I think we need a note for it. From sole user perspective, the conn_type for BigQuery changes from `Google Cloud` -> `Google BigQuery`. However, if we don't change it the current master branch will break or make it impossible on things that rely on `Connection.get_hook` to return a `BigQueryHook`.

Should I update the providers' version by 0.0.1 and add a note to the `ADDITIONAL_INFO.md` or other documentation?

kaxil replied:
Yes probably add an Updating.md style note in `ADDITIONAL_INFO.md` with version as header -- same as https://github.com/apache/airflow/blob/master/UPDATING.md

kaxil replied:
or in https://github.com/apache/airflow/blob/master/airflow/providers/google/CHANGELOG.rst actually

potiuk replied:
Just a comment on this one. Yeah. I wanted to avoid this change when we switched to 2.0 (not to add backwards incompatibilities), but this change indeed breaks the compatibility of Google Provider, so we should mention that in the CHANGELOG @kaxil mentioned. I think however, whether this is a good time to "merge" this change now - we are planning a bigger release of Google Provider (3.0.0) around March/April when we will drop "backports" altogether. We need to update dependencies and upgrade a few remaining APIs of python libraries of Google - and such a backwards-incompatible change could be added then. 

I think if we could make this change without this breaking compatibility now, that would be ideal.

xinbinhuang replied:
@potiuk I want to clarify a bit about the change here. The current Goold Providers in the master branch is broken because it doesn't return  `BigQueryHook` when we do `Connection.get_hook` (this is the original behavior). 

Before these 2 PRs #12466 #12558, `Connection.get_hook` was using a static `conn_type` -> `HookClass` mapping to get the underlying hook. However, after #12558, `airflow.providers.google.common.hooks.base_google.GoogleBaseHook` was added after `BigQueryHook` so `BigQueryHook` will be overridden during providers discovery. https://github.com/PolideaInternal/airflow/blob/802159767baf1768d92c6047c2fdb2094ee7a2a8/airflow/providers/google/provider.yaml#L714-L720

To solve this, I think we can either:
- Fix the bug by putting  `BigQueryHook`  after `GoogleBaseHook`. This will give back us the old behavior.
- Rename the `conn_type` as in this PR, which will involve some versioning consideration.

I think we can fix the bug in this PR, and create a new issue/PR with the new `conn_type`. I am not familiar with the overall releasing schedule and planning, so please let me know which one do you prefer? @potiuk @kaxil

xinbinhuang replied:
I have reverted the `BigQueryHook` conn_type change and put `BigQueryHook` after `GoogleBaseHook` for now.

xinbinhuang committed:
Rename hook_name="Google Bigquery"

xinbinhuang committed:
Revert BigQueryHook change & fix hook discovery

xinbinhuang committed:
Increase druid provider minor version for deprecation

xinbinhuang committed:
fixup! Revert BigQueryHook change & fix hook discovery

kaxil commented on airflow/operators/sql.py:
```diff
@@ -120,16 +148,16 @@ def _convert_to_float_if_possible(s):
 
 
-class SQLValueCheckOperator(BaseOperator):
+class SQLValueCheckOperator(BaseSQLOperator):
```
Since this operator is in Airflow Core, can we release providers with the changes in this PR without corresponding core changes

kaxil replied:
cc @potiuk

xinbinhuang replied:
I think so, the refactor on this operator is mainly to reduce duplicated code, and the logic, interface, and signature don't change in a backward incompatible one. But I don't know enough about how providers are released and how the core may affect it. cc @potiuk (sorry to ping you again..)

kaxil replied:
I think it should work
